include/kokkos-utils/callbacks/TimerListener.hpp

```diff
@@ -0,0 +1,113 @@
+#ifndef KOKKOS_UTILS_CALLBACKS__TIMERLISTENER_HPP
+#define KOKKOS_UTILS_CALLBACKS__TIMERLISTENER_HPP
+
+#include "Kokkos_NumericTraits.hpp"
+
+#include "kokkos-utils/callbacks/Listener.hpp"
+#include "kokkos-utils/callbacks/Matcher.hpp"
+#include "kokkos-utils/concepts/ExecutionSpace.hpp"
+
+namespace Kokkos::utils::callbacks
+{
+
+// template <typename...>
+// class TimerListener;
+
+template <typename TimerType, typename EventType>
+concept TimerStartFor = Event<EventType> && requires (TimerType& timer, const EventType& event) {
+    { timer.start(event) } -> std::same_as<void>;
+};
+
+template <typename TimerType, typename EventType>
+concept TimerStopFor = Event<EventType> && requires (TimerType& timer, const EventType& event) {
+    { timer.stop(event) } -> std::same_as<void>;
+};
+
+/**
+ * @brief Listener that starts a timer when a begin event is received and stops it
+ *        when a corresponding end event is received.
+ *
+ * The start event is selected by requiring it to be of type @p BeginEventType and
+ * to be a match for the passed matcher. The stop event must be of type @p EndEventType
+ * and have the same ID as the begin event.
+ *
+ * The underlying timer @ref timer is a @ref Kokkos::utils::timer::Timer, instantiated
+ * for the execution space @p Exec. This class stores the execution space
+ * instance passed to the constructor, and uses it as the one passed to @ref timer.
+ *
+ * This class is well-suited for timing the execution of a parallel workload, such as a
+ * @c Kokkos parallel for, parallel reduce, parallel scan, or deep copy operation, to which
+ * an execution space instance is passed. In such a situation, the execution space instance
+ * passed to this class's constructor should be the same as the one passed to the workload.
+
+ TODO doc refactor
+ */
+template <
+    Matcher BeginEndMatcherType,
+    typename TimerType
+>
```

Comment:
If the `BeginEndMatcherType` stores the begin and end events, you can check where with a `requires` that the timer can be started and stopped with those event types. Those checks can then be deleted from the body of the function.

Reply:
The current design wouldn't allow such a thing.

For instance, it's allowed you do:
```c++
using my_timer_t = TimerListener<EventRegexMatcher, TimerType>;
```
and `EventRegexMatcher` hasn't any "begin" and "end" event type.

Reply:
Indeed. Got it!